intermediate_source/transformer_building_blocks.py

```diff
@@ -0,0 +1,753 @@
+        self.w3 = nn.Linear(dim, hidden_dim, bias=False, **factory_kwargs)
+    
+    def forward(self, x):
+        return self.w2(F.silu(self.w1(x)) * self.w3(x))
+
+########################################################################
+# An alternative way of implementing this that uses packed projection is 
+
+class PackedSwiGLUFFN(nn.Module):
+    def __init__(self, dim, hidden_dim, multiple_of, ffn_dim_multiplier=None, device=None, dtype=None):
+        factory_kwargs = {'device': device, 'dtype': dtype}
+        super().__init__()
+        hidden_dim = int(2 * hidden_dim / 3)
+        # custom dim factor multiplier
+        if ffn_dim_multiplier is not None:
+            hidden_dim = int(ffn_dim_multiplier * hidden_dim)
+        hidden_dim = multiple_of * ((hidden_dim + multiple_of - 1) // multiple_of)
+
+        self.w13 = nn.Linear(dim, 2 * hidden_dim, bias=False, **factory_kwargs)
+        self.w2 = nn.Linear(hidden_dim, dim, bias=False, **factory_kwargs)
+    
+    def forward(self, x):
+        x1, x3 = torch.chunk(self.w13(x), 2, dim=-1)
+        return self.w2(F.silu(x1) * x3)
+
+################################################################################
+# We can compare the performance of the two implementations as follows
+# Depending on your hardware, you might see different results. On an A100 I see
+# 1.12x speedup for D=128.
+D = 128
+
+swigluffn = torch.compile(SwiGLUFFN(D, D * 4, 256, device='cuda', dtype=torch.bfloat16))
+packed_swigluffn = torch.compile(PackedSwiGLUFFN(D, D * 4, 256, device='cuda', dtype=torch.bfloat16))
+
+q, _, _, sentence_lengths = gen_batch(D, D, D, D, device="cuda", dtype=torch.bfloat16)
+
+# warmup
+swigluffn(q)
+packed_swigluffn(q)
+
+# benchmark
+_, time, _ = benchmark(swigluffn, q)
+_, time_packed, _ = benchmark(packed_swigluffn, q)
+# On my A100 prints 1.08x speedup
+print(f"SwiGLUFFN: {time} s, PackedSwiGLUFFN: {time_packed} s, speedup: {time/time_packed:.2f}x")
+
+################################################################################
+# Extended examples
+# -----------------
+# 
+# We intend to update this tutorial to demonstrate more examples of how to use
+# the various performant building blocks such as KV-Caching, Grouped Query Attention
+# etc. Further, there are several good examples of using various performant building blocks to
+# implement various transformer architectures. Some examples include
+#
+# * `gpt-fast <https://github.com/pytorch-labs/gpt-fast>`_
+# * `segment-anything-fast <https://github.com/pytorch-labs/segment-anything-fast>`_
+# * `lucidrains implementation of NaViT with nested tensors <https://github.com/lucidrains/vit-pytorch/blob/73199ab486e0fad9eced2e3350a11681db08b61b/vit_pytorch/na_vit_nested_tensor.py>`_
+# * `torchtune's implementation of VisionTransformer <https://github.com/pytorch/torchtune/blob/a8a64ec6a99a6ea2be4fdaf0cd5797b03a2567cf/torchtune/modules/vision_transformer.py#L16>`_
```

Comment:
Can we please add a conclusion? Like:

Conclusion
------------

In this tutorial, we have ....